src/ripple/overlay/impl/PeerReservationTable.cpp

```diff
@@ -0,0 +1,168 @@
+//------------------------------------------------------------------------------
+/*
+    This file is part of rippled: https://github.com/ripple/rippled
+    Copyright (c) 2019 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#include <ripple/overlay/PeerReservationTable.h>
+
+#include <ripple/basics/Log.h>
+#include <ripple/core/DatabaseCon.h>
+#include <ripple/json/json_value.h>
+#include <ripple/protocol/PublicKey.h>
+#include <ripple/protocol/jss.h>
+
+#include <boost/optional.hpp>
+
+#include <algorithm>
+#include <iterator>
+#include <mutex>
+#include <string>
+#include <vector>
+
+namespace ripple {
+
+auto
+PeerReservation::toJson() const -> Json::Value
+{
+    Json::Value result{Json::objectValue};
+    result[jss::node] = toBase58(TokenType::NodePublic, nodeId);
+    if (!description.empty())
+    {
+        result[jss::description] = description;
+    }
+    return result;
+}
+
+auto
+PeerReservationTable::list() const -> std::vector<PeerReservation>
+{
+    std::vector<PeerReservation> list;
+    list.reserve(table_.size());
```

Comment:
I understand why you want to do this, but it's not safe to call `table_.size()` outside of the lock.  So you should move this line of code so it is inside the `lock_guard`.

Reply:
Can you teach me why it's not safe?

Reply:
From a standees perspective, the `size()` of `table_` is not synchronized.  Any unsynchronized access invokes undefined behavior.

From a practical perspective, you have two concerns: since the return value of `size()` is bigger than a byte, the value might be torn in which case who knows what value `size()` will return.  Another concern is what the optimizer might do with this stuff.  Who know where `size()` might actually be invoked if you don't place it under the lock.  Of the optimizer might recognize that you have undefined behavior here and remove chunks of the code.

But given what optimizers do today I take any threat of undefined behavior very seriously.

Reply:
An unsynchronized call of a const method is UB? Where can I read more about this. First I've heard of it.

Reply:
TIL any unsynchronized access (even single-writer) of a "memory location" (even a scalar type like `std::size_t`) from multiple threads is UB.

> A memory location is
>
> - an object of scalar type (arithmetic type, pointer type, enumeration type, or `std::nullptr_t`)
> - or the largest contiguous sequence of bit fields of non-zero length

> When an evaluation of an expression writes to a memory location and another evaluation reads or modifies the same memory location, the expressions are said to conflict. A program that has two conflicting evaluations has a data race unless
>
> - both evaluations execute on the same thread or in the same signal handler, or
> - both conflicting evaluations are atomic operations (see std::atomic), or
> - one of the conflicting evaluations happens-before another
>
> If a data race occurs, the behavior of the program is undefined.

https://en.cppreference.com/w/cpp/language/memory_model

Reply:
In http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2019/n4820.pdf (the most current working draft) `Section 6.8.2.1 Data races` paragraph 21.2

> The execution of a program contains a data race if it contains two potentially concurrent conflicting actions, at least one of which is not atomic, and neither happens before the other, except for the special case for signal handlers described below. Any such data race results in undefined behavior. [Note: It can be shown that programs that correctly use mutexes and memory_order::seq_cst operations to prevent all data races and use no other synchronization operations behave as if the operations executed by their constituent threads were simply interleaved, with each value computation of an object being taken from the last side effect on that object in that interleaving. This is normally referred to as “sequential consistency”. However, this applies only to data-race-free programs, and data-race-free programs cannot observe most program transformations that do not change single-threaded program semantics. In fact, most single-threaded program transformations continue to be allowed, since any program that behaves differently as a result must perform an undefined operation.

And this isn't new.  I'm fairly confident it has been true since C++11 when the standard became thread aware.

Reply:
I'm not saying it is new, it's just new to me. I stepped away from C++ soon after C++11 was passed, and I was never exposed to the concurrency model. I'm still catching up.